File: seminar-10/kosaraju/cpp/kosaraju.cpp
#include <algorithm>
#include <vector>

using namespace std;

void dfs1(int v, const vector<vector<int>> &graph, vector<bool> &used, vector<int> &order) {
    used[v] = true;
    for (int to : graph[v]) {
        if (!used[to]) {
            dfs1(to, graph, used, order);
        }
    }
    order.push_back(v);
}

void dfs2(int v, const vector<vector<int>> &reversed_graph, vector<bool> &used, vector<int> &component, int color) {
    used[v] = true;
    component[v] = color;
    for (int to : reversed_graph[v]) {
        if (!used[to]) {
            dfs2(to, reversed_graph, used, component, color);
        }
    }
}

vector<int> kosaraju(const vector<vector<int>> &graph) {
    int n = (int) graph.size();
    vector<vector<int>> reversed_graph(n);

    for (int v = 0; v < n; v++) {
        for (int to : graph[v]) {
            reversed_graph[to].push_back(v);
        }
    }

    vector<bool> used(n, false);
    vector<int> order;

    for (int v = 0; v < n; v++) {
        if (!used[v]) {
            dfs1(v, graph, used, order);
        }
    }

    reverse(order.begin(), order.end());

    fill(used.begin(), used.end(), false);
    vector<int> component(n, -1);
    int color = 0;

    for (int v : order) {
        if (!used[v]) {
            dfs2(v, reversed_graph, used, component, color);
            color++;
        }
    }

    return component;
}

File: seminar-10/kosaraju/python/kosaraju.py
def dfs1(v, graph, used, order):
    used[v] = True
    for to in graph[v]:
        if not used[to]:
            dfs1(to, graph, used, order)
    order.append(v)


def dfs2(v, reversed_graph, used, component, color):
    used[v] = True
    component[v] = color
    for to in reversed_graph[v]:
        if not used[to]:
            dfs2(to, reversed_graph, used, component, color)


def kosaraju(graph):
    """
    graph — список смежности ориентированного графа
    возвращает массив component, где component[v] — номер КСС вершины v
    """
    n = len(graph)
    reversed_graph = [[] for _ in range(n)]

    for v in range(n):
        for to in graph[v]:
            reversed_graph[to].append(v)

    used = [False] * n
    order = []

    for v in range(n):
        if not used[v]:
            dfs1(v, graph, used, order)

    order.reverse()

    used = [False] * n
    component = [-1] * n
    color = 0

    for v in order:
        if not used[v]:
            dfs2(v, reversed_graph, used, component, color)
            color += 1

    return component
